### PR TITLE
Fix OCSP nonce handling

### LINKED ISSUE #26439: Incorrect position of nonce in OCSP request
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

authentication

### Describe the bug

The nonce of the OCSP request sent by Keycloak is different from RFC 6960. 
According to RFC 6960, nonces should be sent as requestExtensions, but Keycloak does not.

https://datatracker.ietf.org/doc/html/rfc6960#section-4.1
https://datatracker.ietf.org/doc/html/rfc6960#section-4.4.1 

Therefore, for example, when using OpenSSL as an OCSP responder, 
the nonce is not recognized correctly and the nonce value is not returned in the OCSP response.

Below is the location of the nonce when capturing OCSP request packets using Keycloak and OpenSSL.

Case: Keycloak OCSP request(different from RFC6960)
----------------------------------------------------------------------------
- tbsRequest
  - requestList
    - reqCert
    - singleRequestExtensions
      - Extension(nonce) <-- NG

Case: OpenSSL OCSP client(matches RFC6960)
----------------------------------------------------------------------------
- tbsRequest
  - requestList
    - reqCert
  - requestExtensions
    - Extension(nonce) <-- OK

### Version

23.0.4

### Expected behavior

OCSP response includes nonce

### Actual behavior

OCSP response does not include nonce

### How to Reproduce?

1. Config x509 authentication to the authentication flow according to https://www.keycloak.org/docs/latest/server_admin/index.html#_x509.
2. Set "OCSP Checking Enabled" to "ON" and "OCSP Responder URI" to "http://localhost:8088"
3. Start OCSP responder on port 8088 with openssl
`
openssl ocsp -index {index.txt file path} -rsigner {CA certificate path} -rkey {CA private key path} -port 8088 -CA {CA certificate path} -text
`
4. Execute X509 certificate authentication
5. If you capture a packet from port 8088, you will see that no nonce is being sent in the OCSP response.

### Anything else?

In addition to the above, when looking at the OCSP request with wireshark, it appears that only the nonce value is not output correctly and is recognized as an invalid packet. There seems to be another problem with how the nonce value is stored.

Also, related to this issue, the nonce check is not performed if the OCSP response does not contain a nonce, so it is difficult to notice the problem of missing nonce. You may need to consider making nonce checking mandatory as an option, or issuing a warning log.

### PR DESCRIPTION
**Pull request by justin-tay**
Closes #26439

This fixes the issue with the OCSP nonce as setting with `setRequestExtensions` sets it properly in `requestExtensions` while setting it as the second parameter to `addRequest` actually sets it in `singleRequestExtensions`.

For testing the `OcspHandler` was modified to require that the nonce is present otherwise it will return a `malformedRequest`.

Debug logs were added if no OCSP nonce is found in the response.

### CODE REVIEW

**Review by rmartinc — CHANGES_REQUESTED** [review]
Thanks @justin-tay for the PR!

I think the changes in the PR are OK. But I also detected that the nonce that is added right now is weird. At least wireshark detected the extension as incorrect saying something about not proper encoding. Reading about this I detected that the `nonce` extension is recommended to be encapsulated in another octet string (see this [issue](https://github.com/pyca/cryptography/issues/6404) for example). I checked openssl and it performs that. Wireshark does not show any error with the double octet string either. I also saw that this example [here](https://www.ietf.org/id/draft-ietf-lamps-ocsp-nonce-update-06.html) also performs double octet string. So I have added that change too over you PR. Besides I changed to create a nonce of 16 bytes because before I saw a nonce of 4 bytes which are too short IMHO.

In summary, can you please add this commit https://github.com/rmartinc/keycloak/commit/4f828518f6e3bfef8dc9ba215bb2f7a10c4d2a36 in the PR branch and check in your environment? If it works for you too just add it to this PR (you can squash it, modify or whatever you want to do with it).

Thanks again!

**Comment by justin-tay** [comment]
Thanks a lot for the review and the fixes @rmartinc !

I've tested the changes and it works well. Have squashed the commit. Thanks!

**Review by rmartinc — APPROVED** [review]
Thanks a lot @justin-tay for this! Approved!
@mposolda This one is ready to go from my point of view.

**Review by mposolda — APPROVED** [review]
@justin-tay @rmartinc Nice, Thanks!
